### ISSUE TITLE
Add more base archival/recall memory stores

### ISSUE THREAD

**Issue by cpacker** [cpacker/MemGPT#32]
TODO: add examples for backing archival/recall memory with real databases:

- MongoDB
- Some vector DB (chroma, etc)

---

### Create your own `PersistenceManager` subclass

https://github.com/cpacker/MemGPT/blob/main/memgpt/persistence_manager.py#L31-L41

The init of this persistence manager should do the appropriate initialization of your database. For example, initializing a connection to it. It also needs to implement utility methods like trimming etc.

### Create your own database-backed `Archival` and `Recall` memory stores

These need to support the insert and query operations that MemGPT will attempt to use in `agent.py`

---

### Top-level PersistenceManager (each agent gets their own)

```python
class DatabasePersistenceManager(PersistenceManager):
    """Sync Agent with database on every action/event"""

    def __init__(self, user_information, database_information):

        # initialize some abstraction over the database (optional)
        self.db_handler = db_handler

        # set up the PersistenceManager to point at this particular agent/user data in the database

    def init(self, agent):

        # create the user/agent in the database

        # also initialize the underlying data stores (note: you can have more than just these two, but you then need to modify MemGPT function schema + system instructions to match)
        self.recall_memory = DatabaseRecallMemory(...)
        self.archival_memory = DatabaseArchivalMemory(...)

    def trim_messages(self, num):
        raise NotImplementedError

    def prepend_to_messages(self, added_messages):
        raise NotImplementedError

    def append_to_messages(self, added_messages):
        raise NotImplementedError

    def swap_system_message(self, new_system_message):
        raise NotImplementedError

    def update_memory(self, new_memory):
        raise NotImplementedError
```

### Individual Archival and Recall memory stores

```python
class DatabaseArchivalMemory(ArchivalMemory):

    def __init__(self, user_information, database_information):
        # initialize the connection to the database that backs Archival

    async def insert(self, memory_string):
        if not isinstance(memory_string, str):
            return
        inserted_id = await self.db_handler.create_archival_memory(...)  # this inserts into the real database 
        return inserted_id

    async def search(self, query_string, count=None, start=None):
        results, total = await self.db_handler.search_archival_memory(...)  # this searches the real database

        # agent call expects: {
        #   'timestamp': ...,
        #   'message': {
        #       'role': ...,
        #       'content' ...,
        #   } 
        # }
        parsed_results = ...
        return parsed_results, total 

class DatabaseRecallMemory(RecallMemory):

    def __init__(self, db_handler, agent_id, api_key):
        # initialize the connection to the database that backs Recall (conversation history)

    async def search(self, query_string, count=None, start=None):
        results, total = self.db_handler.search_messages_text_query_paged(...)  # can be strict or fuzzy text search, or embedding search, or whatever you want

        # agent call expects: {
        #   'timestamp': ...,
        #   'message': {
        #       'role': ...,
        #       'content' ...,
        #   } 
        # }
        parsed_results = ...
        return parsed_results, total 

    async def date_search(self, start_date, end_date, count=None, start=None):
        results, total = self.db_handler.search_messages_by_date_paged(...)  # this does data-based search (optional)

        parsed_results = ...
        return parsed_results, total 
```

**Comment by danx0r** [comment]
I have extensive experience with MongoDB, and I think it would be an excellent choice for a unified persistent db. I see some work has been done in a similar direction with Postgres. Would it be useful for me to implement a mongDB option?

**Comment by sarahwooders** [comment]
@danx0r yes that would be great! If you could follow the example with Postgres that was recently merged that would be great. And feel free to submit a draft pull request for feedback.

**Comment by sahusiddharth** [comment]
Greetings, @cpacker and  @sarahwooders 

I would love to work on this, there are a few things I wanted to discuss, sent you LinedIn connection request with the name Siddharth Sahu.

**Comment by sarahwooders** [comment]
Closed via #708
